### PR TITLE
perform cleanup in batch

### LINKED ISSUE #8: Outbox cleanup query uses unbounded DELETE
Currently the Outbox cleaner deletes all the dispatched Outbox records using a single query, regardless of how many records there are. For endpoints with heavy traffic (a lot of new records inserted in a short period of time) this will lead to excessive locking (page locks) and finally to deadlocks and endpoint shutdowns.

Removing many records in one transaction is also quite a heavy operation for the transaction log.

The suggested solution is to use a loop with batched deletes (each in a separate transaction), very much like here:

https://github.com/Particular/NServiceBus.SqlServer/blob/develop/src/NServiceBus.SqlServer/Receiving/ExpiredMessagesPurger.cs#L35-L41

### PR DESCRIPTION
**Pull request by SimonCropp**
fixes #8

### CODE REVIEW

**Comment by SzymonPobiega** [comment]
Ready for review?

**Comment by SimonCropp** [comment]
Let me add a test first

**Comment by SimonCropp** [comment]
@SzymonPobiega ready for a review

**Comment by SimonCropp** [comment]
or @weralabaj

File: src/SqlPersistence/Outbox/OutboxCommandBuilder.cs [hunk 1]
```diff
@@ -62,5 +61,24 @@ public static OutboxCommands Build(string tablePrefix, string schema, SqlVariant
...
         }
 
+        static string GetCleanupCommand(SqlVariant sqlVariant, string tableName)
+        {
+            switch (sqlVariant)
+            {
+                case SqlVariant.MsSqlServer:
+                    return $@"
+delete top (@BatchSize) from {tableName}
+where Dispatched = 1
```
Comment (weralabaj) [inline src/SqlPersistence/Outbox/OutboxCommandBuilder.cs]: > where Dispatched = 1

why not `=true` anymore?

File: src/SqlPersistence/Outbox/OutboxPersister.cs [hunk 2]
```diff
@@ -15,11 +15,13 @@
 class OutboxPersister : IOutboxStorage
 {
     Func<DbConnection> connectionBuilder;
+    int cleanupBatchCount;
     OutboxCommands outboxCommands;
 
-    public OutboxPersister(Func<DbConnection> connectionBuilder, string tablePrefix, string schema, SqlVariant sqlVariant)
+    public OutboxPersister(Func<DbConnection> connectionBuilder, string tablePrefix, string schema, SqlVariant sqlVariant, int cleanupBatchCount = 10000)
```
Comment (weralabaj) [inline src/SqlPersistence/Outbox/OutboxPersister.cs]: > cleanupBatchCount 

IMO cleanupBatchSize is a better name, as count might indicate how many batches we want to divide that into

Comment (weralabaj) [inline src/SqlPersistence/Outbox/OutboxPersister.cs]: is 10000 just random magic number?

**Comment by weralabaj** [comment]
@SimonCropp few small questions/comments, overall LGTM

File: src/SqlPersistence/Outbox/OutboxCommandBuilder.cs (hunk 1 above)
Comment (SimonCropp) [inline src/SqlPersistence/Outbox/OutboxCommandBuilder.cs]: fixed

File: src/SqlPersistence/Outbox/OutboxPersister.cs (hunk 2 above)
Comment (SimonCropp) [inline src/SqlPersistence/Outbox/OutboxPersister.cs]: stolen from the sql transport. yes magic number

Comment (SimonCropp) [inline src/SqlPersistence/Outbox/OutboxPersister.cs]: cleanupBatchSize is better. fixed

**Pull request merged**